controllers/space/space_controller.go

```diff
@@ -0,0 +1,340 @@
+		return nil
+	}
+
+	return r.updateStatus(space, nsTmplSetReady) // replicates the NSTemplateSet's `ready` condition into the Space, including when `ready/true/provisioned`
+}
+
+func (r *Reconciler) newNSTemplateSet(memberOperatorNS string, space *toolchainv1alpha1.Space) (*toolchainv1alpha1.NSTemplateSet, error) {
+	// look-up the NSTemplateTier
+	tmplTier := &toolchainv1alpha1.NSTemplateTier{}
+	if err := r.Client.Get(context.TODO(), types.NamespacedName{
+		Namespace: space.Namespace,
+		Name:      space.Spec.TierName,
+	}, tmplTier); err != nil {
+		return nil, err
+	}
+	// create the NSTemplateSet from the NSTemplateTier
+	nsTmplSet := &toolchainv1alpha1.NSTemplateSet{
+		ObjectMeta: metav1.ObjectMeta{
+			Namespace: memberOperatorNS,
+			Name:      space.Name,
+		},
+		Spec: *usersignup.NewNSTemplateSetSpec(tmplTier),
+	}
+	return nsTmplSet, nil
+}
+
+func (r *Reconciler) ensureSpaceDeletion(logger logr.Logger, space *toolchainv1alpha1.Space) error {
+	logger.Info("terminating Space")
+	if isBeingDeleted, err := r.deleteNSTemplateSet(logger, space); err != nil {
+		logger.Error(err, "failed to delete the NSTemplateSet")
+		return r.setStatusTerminatingFailed(logger, space, err)
+	} else if isBeingDeleted {
+		if err := r.setStatusTerminating(space); err != nil {
+			logger.Error(err, "error updating status")
+			return err
+		}
+		return nil
+	}
+	// Remove finalizer from Space
+	util.RemoveFinalizer(space, toolchainv1alpha1.FinalizerName)
+	if err := r.Client.Update(context.TODO(), space); err != nil {
+		logger.Error(err, "failed to remove finalizer")
+		return r.setStatusTerminatingFailed(logger, space, err)
+	}
+	logger.Info("removed finalizer")
+	// no need to update the status of the Space once the finalizer has been removed, since
+	// the resource will be deleted
+	return nil
+}
+
+// deleteNSTemplateSet triggers the deletion of the NSTemplateSet on the target member cluster.
+// Returns `true/nil` if the NSTemplateSet is being deleted (whether deletion was triggered during this call,
+// or if it was triggered earlier and is still in progress)
+// Returns `false/nil` if the NSTemplateSet doesn't exist anymore,
+//   or if there is no target cluster specified in the given space, or if the target cluster is unknown.
+// Returns `false/error` if an error occurred
+func (r *Reconciler) deleteNSTemplateSet(logger logr.Logger, space *toolchainv1alpha1.Space) (bool, error) {
+	targetCluster := space.Spec.TargetCluster
+	if targetCluster == "" {
+		targetCluster = space.Status.TargetCluster
+	}
+	if targetCluster == "" {
+		logger.Info("cannot delete NSTemplateSet: no target cluster specified")
+		return false, nil // skip NSTemplateSet deletion
+	}
+	memberCluster, found := r.MemberClusters[targetCluster]
+	if !found {
+		logger.WithValues("target_cluster", targetCluster).Info("Cannot delete NSTemplateSet: unknown target member cluster")
+		return false, nil // skip NSTemplateSet deletion
```

Comment:
If the targetCluster is set but is not found, then this is a problem - we should return an error here!

Reply:
you do realize that you suggested this code in https://github.com/codeready-toolchain/host-operator/pull/548/commits/2fcdba0b8ab1599fa4e03b052b39a8d59ffd771d?

Reply:
I'm fine with returning an error, but then what are we supposed to do with the Space resource? Leave it in a `TerminatingFailed` state? Let the reconciler return the error until the Space resource has a `targetCluster` set by an admin?

Reply:
> you do realize that you suggested this code in 2fcdba0?

no, my suggestion didn't contain the "return" line - this is what I suggested
```go 
	targetCluster := space.Spec.TargetCluster
	if targetCluster == "" {
		targetCluster = space.Status.TargetCluster
	}
	if targetCluster == "" {
		logger.Info("cannot delete NSTemplateSet: no target cluster specified")
		return false, nil // skip NSTemplateSet deletion
	}
	memberCluster, found := r.MemberClusters[targetCluster]
	if !found {
		logger.WithValues("target_cluster", targetCluster).Info("Cannot delete NSTemplateSet: unknown target member cluster")
```

> I'm fine with returning an error, but then what are we supposed to do with the Space resource? Leave it in a TerminatingFailed state? Let the reconciler return the error until the Space resource has a targetCluster set by an admin?

yes, exactly - keep it in the failed state. The reason can be either a wrong targetCluster name or missing ToolchainCluster CR when the operator deployment was started.

Reply:
> no, my suggestion didn't contain the "return" line - this is what I suggested

ok, I'm confused because what you said you suggested contains exactly the same code:
```
return false, nil // skip NSTemplateSet deletion
```

that being said, ok, let me return an error, and we'll see how we can deal with this case when/if it happens

Reply:
ok, I see. We have 2 identical lines like this
```
return false, nil // skip NSTemplateSet deletion
``` 
and I was looking at the wrong case :(